src/Symfony/Component/Security/Http/Authenticator/Passport/Credentials/PasswordCredentials.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\Authenticator\Passport\Credentials;
+
+/**
+ * Implements password credentials.
+ *
+ * These plaintext passwords are checked by the UserPasswordEncoder during
+ * authentication.
+ *
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @final
+ * @experimental in 5.1
+ */
+class PasswordCredentials implements CredentialsInterface
+{
+    private $password;
+    private $resolved = false;
+
+    public function __construct(string $password)
+    {
+        $this->password = $password;
+    }
+
+    public function getPassword(): string
+    {
+        return $this->password;
+    }
+
+    /**
+     * @internal
+     */
+    public function markResolved(): void
+    {
+        $this->resolved = true;
```

Comment:
This may be an over-optimization for security, but for this badge and also `PasswordUpgradeBadge`, I wondered if, when marking it as resolved, the "plain text password" should be nullified (and an exception thrown if someone tries to access it later). That could be overkill. I know we have `UserInterface::eraseCredentials()`, but the `UserInterface` is eventually serialized to the session - so that is extra sensitive.